File: crystal/tests/primitive.cc
#include "LaDaConfig.h"
#include<iostream>
#include<string>
#include<vector>

#include "../supercell.h"
#include "../primitive.h"

using namespace std;
int main()
{
  using namespace LaDa;
  using namespace LaDa::crystal;
  using namespace LaDa::math;
  typedef TemplateStructure< std::vector<std::string> > t_Str; 
  TemplateStructure< std::vector<std::string> > lattice; 
  lattice.set_cell(0,0.5,0.5)
                  (0.5,0,0.5)
                  (0.5,0.5,0);
  lattice.add_atom(0,0,0, "Si")
                  (0.25,0.25,0.25, "Si", "Ge");
  rMatrix3d cell;

  size_t nmax= LADA_LIM;
  bool cont = true;
  for(size_t a(1); a <= nmax and cont; ++a)
  {
    if(nmax % a != 0) continue;
    size_t const Ndiv_a = nmax/a;
    cell(0,0) = a;
    // Iterates over values of a such that a * b * c == nmax
    for(size_t b(1); b <= Ndiv_a and cont; ++b) 
    {
      cell(1,1) = b;
      if(Ndiv_a % b != 0) continue;
      // Iterates over values of a such that a * b * c == nmax
      size_t const c( Ndiv_a/b);
      cell(2,2) = c;
      if( a * b *c != nmax ) BOOST_THROW_EXCEPTION(LaDa::error::internal());
      for(size_t d(0); d < b and cont; ++d) 
      {
        cell(1,0) = d;
        for(size_t e(0); e < c and cont; ++e) 
        {
          cell(2,0) = e;
          for(size_t f(0); f < c and cont; ++f) 
          {
            cell(2,1) = f;
            TemplateStructure< std::vector<std::string> > structure = supercell(lattice, cell);
            structure = primitive(structure);
            if( not eq(lattice.cell().determinant(), structure.cell().determinant(), 1e-5) )
            {
              std::cout << cell << "\n\n";
              std::cout << structure.cell() << "\n\n";
              std::cout << lattice.cell().inverse() * structure.cell() << "\n\n";
            }
            LADA_DOASSERT(eq(lattice.cell().determinant(), structure.cell().determinant(), 1e-5), "Not primitive.\n");
            LADA_DOASSERT(is_integer(structure.cell() * lattice.cell().inverse(), 1e-5), "Not a sublattice.\n");
            LADA_DOASSERT(is_integer(lattice.cell() * structure.cell().inverse(), 1e-5), "Not a sublattice.\n");
            t_Str::const_iterator i_atom = structure.begin();
            t_Str::const_iterator const i_atom_end = structure.end();
            for(; i_atom != i_atom_end; ++i_atom)
            {
              LADA_DOASSERT(compare_sites(lattice[i_atom->site])(i_atom->type), "Inequivalent occupation.\n");
              LADA_DOASSERT( is_integer(lattice.cell().inverse()*(i_atom->pos - lattice[i_atom->site].pos), 1e-5),
                             "Inequivalent positions.\n") 
            }
          } // f
        } // e
      } // d
    } // b
  } // a

  return 0;
}

File: crystal/tests/primitive.py
""" Tests for primitive and is_primitive. """

def itercells(nmax):
  """ Iterates over all cells with ``nmax`` and fewer atoms. """
  from numpy import zeros
  cell = zeros((3,3), dtype="float64")
  # loop over all possible cells with less than 10 atoms. 
  for n in xrange(2, nmax+1):
    for a in xrange(1, n+1):
      if n % a != 0: continue;
      ndiv_a = n / a
      cell[0,0] = a
      for b in xrange(1, ndiv_a+1):
        if ndiv_a % b != 0: continue
        cell[1,1] = b
        c = ndiv_a / b
        cell[2,2] = c
        for d in xrange(b):
          cell[1,0] = d
          for e in xrange(c):
            cell[2,0] = e
            for f in xrange(c):
              cell[2,1] = f
              yield cell

def test_primitive():
  """ Tests whether primitivization works. """
  from numpy import abs, dot
  from numpy.linalg import inv
  from pylada.crystal.cppwrappers import supercell, Structure, are_periodic_images as api, \
                                       primitive, is_primitive
  from pylada.math import is_integer

  lattice = Structure( 0.0, 0.5, 0.5,
                       0.5, 0.0, 0.5,
                       0.5, 0.5, 0.0, scale=2.0, m=True ) \
                     .add_atom(0, 0, 0, "As")           \
                     .add_atom(0.25, 0.25, 0.25, ['In', 'Ga'], m = True)
  assert is_primitive(lattice)
  for cell in itercells(10): 
    structure = supercell(lattice, dot(lattice.cell, cell))
    assert not is_primitive(structure)
#   structure = primitive(structure, 1e-8)
#   assert is_primitive(structure)
#   assert abs(structure.volume - lattice.volume) < 1e-8
#   assert len(structure) == len(lattice)
#   assert is_integer(dot(structure.cell, inv(lattice.cell)))
#   assert is_integer(dot(lattice.cell, inv(structure.cell)))
#   invcell = inv(lattice.cell)
#   for atom in structure:
#     assert api(lattice[atom.site].pos, atom.pos, invcell) and \
#            atom.type == lattice[atom.site].type and \
#            getattr(lattice[atom.site], 'm', False) == getattr(atom, 'm', False) and \
#            (getattr(atom, 'm', False) or atom.site == 0)

if __name__ == "__main__":
  from sys import argv, path 
  if len(argv) > 0: path.extend(argv[1:])

  test_primitive()
